src/rules/flake8_return/rules.rs

```diff
@@ -319,10 +330,18 @@ fn unnecessary_assign(checker: &mut Checker, stack: &Stack, expr: &Expr) {
             return;
         }
 
-        checker.diagnostics.push(Diagnostic::new(
-            UnnecessaryAssign,
-            Range::from_located(expr),
-        ));
+        let mut diagnostic = Diagnostic::new(UnnecessaryAssign, Range::from_located(expr));
+
+        if checker.patch(diagnostic.kind.rule()) {
+            if let Some(assign_expr) = stack.assign_values.get(id.as_str()) {
+                diagnostic.amend(Fix::replacement(
+                    unparse_expr(assign_expr, checker.stylist),
+                    expr.location,
+                    expr.end_location.expect("Expression has no end location"),
+                ));
+            }
+        }
+        checker.diagnostics.push(diagnostic);
```

Comment:
I think the framing here should be that we want to replace the `Assign` statement and the `Return` statement with the `return ${value}`.

Something like:

```rust
diagnostic.amend(Fix::replacement(
    unparse_expr(assign_expr, checker.stylist),
    assign_stmt.location,
    return_stmt.end_location.unwrap(),
));
```

But, we should take care that these statements are conse

Reply:
Another thing I'm slightly confused on is this code I'm trying:
```rust
if let Some(assign_loc) = stack.assigns.get(id.as_str()){
    println!("return {:?}", assign_loc.last().unwrap());
    diagnostic.amend(Fix::replacement(
        unparse_expr(assign_expr, checker.stylist),
        *assign_loc.last().unwrap(),
        expr.end_location.unwrap(),
    ));
}
```
This code causes ruff to hang, and the prints show why:
```console
return Location { row: 5, column: 4 }
return Location { row: 12, column: 4 }
return Location { row: 20, column: 4 }
return Location { row: 19, column: 4 }
return Location { row: 21, column: 4 }
return Location { row: 25, column: 4 }
return Location { row: 33, column: 4 }
... (this repeats for a LOT of lines)
```
any clue why this could be happening?

Reply:
I think for now we limit this to the version in which the assignment and the return are consecutive statements, if we can.

If we want to support fixes in which there are interleaving statements, we have to do something clever like structure the fix as a replacement with "all the content before the statement, followed by the new return" (rather than two separate fixes).

Reply:
@charliermarsh 
This is the command:
```console
> cargo run resources/test/fixtures/flake8_return/RET504.py --ignore=ALL --extend-select RET504 --no-cache --fix
```
The only difference in the code is adding the `*assign_loc.last().unwrap(),` and `if let Some(assign_loc) = stack.assigns.get(id.as_str())` with the `println`. I don't see why this would be called multiple times - perhaps I am simply unfamiliar with the `ruff` internals. The location and fix is correct, but then it gets called over and over and gets stuck.

Reply:
I haven't invested whether it handles all the cases we care about, but this version doesn't suffer from whatever recursion issue was being hit previously:

```rust
if checker.patch(diagnostic.kind.rule()) {
    if let Some(assign_expr) = stack.assign_values.get(id.as_str()) {
        if let Some(assign_loc) = stack.assigns.get(id.as_str()) {
            diagnostic.amend(Fix::replacement(
                unparse_stmt(
                    &create_stmt(StmtKind::Return {
                        value: Some(Box::new(assign_expr.clone())),
                    }),
                    checker.stylist,
                ),
                *assign_loc.last().unwrap(),
                expr.end_location.unwrap(),
            ));
        }
    }
}
```

Reply:
As an aside, this:

```py
def f(x):
    val = 2
    print(x)
    return val
```

Actually doesn't trigger `RET504` if I'm not mistaken, since we treat any function calls between the assignment and the return (even if they don't involve the variable) as negating the rule.

Reply:
Ah, I think you're right -- I should have this finished by tomorrow. My only other thought is how to work with the `Assign` statement being on another line, how would you ensure that there is just the assignment + whitespace/comment on the next few lines, would you have loop through the source code directly and check if each line is whitespace/comment or is there a better way?

What I'm talking about is something like this:
```py
def safe_average(lst):
    """There is a extra newline and extra indentation after the assignment"""
    return_value = 2.4
    # This is also a comment explaining why we need to check for comments
    """
    Also, the next comment has *critical* information, but would the autofix remove it?
    I presume we would have to keep track of this, or just ignore this for now and 
    put it in an issue somewhere.
    """
    # We return this function in this manner because if the length was zero we would raise a ZeroDivisionError
    return sum(lst) / len(lst) if len(lst) > 0 else 0 
```

Reply:
So here's what I'd try to do (though it'll require more bookkeeping than we do now)...

When we track the assignment, track the start and end locations of the entire assignment statement.

Then, when we create the fix, structure it as: "Replace all the content from the start of the assignment with: (1) the content from the end of the assignment to the start of the return + (2) the modified return statement."

Reply:
We also have the ability to skip the fix if there's a comment between the assign and the return -- we have a `has_comments_in` utility that takes a range as input, and returns true if there are any comments. So we could do that instead.

Or we could even really restrictive and say that we're only going to do this fix if it's purely whitespace between the assignment and the return. That's actually pretty reasonable IMO!

Reply:
I'll try for the first option, if it seems a bit too difficult (shouldn't be, but will require a bit more restructuring and there's also a whitespace issue, but Black would fix that... not sure what the policy is for that) it's a bit late for me right now so I'll get to it by tomorrow. Sorry if I've asked too many questions or been a hassle, but thanks for helping me.

Reply:
No apology needed at all. I wish I could be even more helpful! Just trying to juggle a few things at once. Appreciate your effort on the issue :)